src/Common/ProxyConfiguration.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <string>
+
+namespace DB
+{
+
+struct ProxyConfiguration
+{
+    enum class Protocol
+    {
+        HTTP,
+        HTTPS,
+        ANY
```

Comment:
That's not the only difference. The main and most important difference is that `ClientConfigurationPerRequest` is a S3 specific structure and relies on `Aws::Http::Scheme` (a third party module enum).

`ProxyConfiguration` is a ClickHouse domain enum, meant to be used across ClickHouse source code. It removes this extra dependency and tight coupling.

For instance, URL functions also need proxy, but should not have a dependency on AWS stuff. There might be other parts of CH that depend on proxy stuff and will adhere to this implementation later, no dependency on AWS stuff.

Reply:
Not only that, but `ProxyConfiguration::Protocol::ANY` is not handled as `HTTP`. Both `RemoteResolver` and `ListResolver` will literally use ANY (HTTP or HTTPs) if ANY is specified. You can check that in `ProxyConfigurationResolverProvider.cpp`.

You might have been confused by the `EnvironmentResolver`, which is not currently handling ANY as ANY. I'll address that.

Reply:
> ClientConfigurationPerRequest is a S3 specific structure and relies on Aws::Http::Scheme (a third party module enum)

I don't mind using `Aws::Http::Scheme` in ClickHouse code (since we already do), but okay

> You might have been confused by the EnvironmentResolver, which is not currently handling ANY as ANY

Yes, and also by
https://github.com/ClickHouse/ClickHouse/blob/63c7ba51a779dd3bb9a7562c47fd6f9f26327f27/src/IO/S3/ProxyConfigurationResolverAdapter.cpp#L18-L20

Reply:
However, if I'm not mistaken, ANY is used only as an argument of a factory (to get any proxy) and is never used as a part of an actual proxy configuration (each proxy is either HTTP or HTTPS). In other words, can `ProxyConfiguration::protocol` be `ANY`?

Reply:
Yes, `ANY` is only used to get `any resolver`, but `ProxyConfiguration::protocol` will never be `ANY`.

The only reason `ANY` exists is to keep backwards compatibility with the "old" behavior, where the user did not have the option to specify if the proxy was meant for HTTP or for HTTPs requests. ClickHouse would pick any.

So `ANY` is only used in the context of the S3 storage conf proxy.

Reply:
> Yes, ANY is only used to get any resolver, but ProxyConfiguration::protocol will never be ANY.

So there's no real difference between `ClientConfigurationPerRequest` and `ProxyConfiguration` (except for using a third party module enum)

Reply:
Yes, you are correct